src/frontend/src/optimizer/rule/apply_join.rs

```diff
@@ -0,0 +1,302 @@
+            .enumerate()
+            .map(|(i, field)| {
+                Self::create_equal_expr(
+                    i,
+                    field.data_type.clone(),
+                    i + join_left_len + apply_left_len,
+                    field.data_type.clone(),
+                )
+            })
+            .collect_vec();
+        let new_join_condition = Condition {
+            conjunctions: join
+                .on()
+                .clone()
+                .into_iter()
+                .map(|expr| rewriter.rewrite_expr(expr))
+                .chain(natural_conjunctions.into_iter())
+                .collect_vec(),
+        };
+
+        let mut left_apply_condition: Vec<ExprImpl> = vec![];
+        let mut right_apply_condition: Vec<ExprImpl> = vec![];
+        let mut other_condition: Vec<ExprImpl> = vec![];
+
+        match join.join_type() {
+            JoinType::LeftSemi | JoinType::LeftAnti => {
+                left_apply_condition.extend(apply_on);
+            }
+            JoinType::RightSemi | JoinType::RightAnti => {
+                right_apply_condition.extend(apply_on);
+            }
+            JoinType::Inner | JoinType::LeftOuter | JoinType::RightOuter | JoinType::FullOuter => {
+                let mut d_t1_bit_set =
+                    FixedBitSet::with_capacity(apply_left_len + join.base.schema.len());
+                let mut d_t2_bit_set =
+                    FixedBitSet::with_capacity(apply_left_len + join.base.schema.len());
+                d_t1_bit_set.set_range(0..apply_left_len + join_left_len, true);
+                d_t2_bit_set.set_range(0..apply_left_len, true);
+                d_t2_bit_set.set_range(
+                    apply_left_len + join_left_len..apply_left_len + join.base.schema.len(),
+                    true,
+                );
+
+                for (key, group) in &apply_on.into_iter().group_by(|expr| {
+                    let mut visitor =
+                        CollectInputRef::with_capacity(apply_left_len + join.base.schema.len());
+                    visitor.visit_expr(expr);
+                    let collect_bit_set = FixedBitSet::from(visitor);
+                    if collect_bit_set.is_subset(&d_t1_bit_set) {
+                        0
+                    } else if collect_bit_set.is_subset(&d_t2_bit_set) {
+                        1
+                    } else {
+                        2
+                    }
+                }) {
+                    let vec = group.collect_vec();
+                    match key {
+                        0 => left_apply_condition.extend(vec),
+                        1 => right_apply_condition.extend(vec),
+                        2 => other_condition.extend(vec),
+                        _ => unreachable!(),
+                    }
+                }
+
+                // rewrite right condition
+                let mut right_apply_condition_rewriter = Rewriter {
+                    join_left_len: apply_left_len,
```

Comment:
`Rewriter.join_left_len` should be `apply_left_len + join_left_len`? Because the original Apply's schema is | D | T1 | T2 |, and `right_condition` refer to cols in T2 whose offset is len of D + len of T1.

Reply:
Oh, use apply_left_len here won't affect correctness?

Reply:
Also thought about this. I think any value between `apply_left_len ` and `apply_left_len + join_left_len` (inclusive on both sides) is correct, as right condition only index into D or T2.